File: 2019W04/PreppindataW4.py
import pandas as pd
import re
from dateutil.relativedelta import relativedelta

xls = pd.ExcelFile("E:\PD - ESPN stats.xlsx")

def clean_WL(wl):
    if wl[0:5]=='2019-':
        wl = re.search('\d+-\d+\s', wl).group(0).strip()
        wl = '-'.join(s.lstrip('0') for s in wl.split('-')[::-1])
    else:
        wl
    return wl    

espn = pd.read_excel(xls,"Sheet1",converters={'W-L':str}) 
espn['W-L'] = espn.apply(lambda x: clean_WL(x['W-L']),axis=1) 


for cols in ['HI POINTS','HI REBOUNDS','HI ASSISTS']:
    temp = espn[cols].str.split(' ', n = 1, expand = True) 
    espn[cols+' - Player'] = temp[0]
    espn[cols+' - Value'] = temp[1].astype('int64')
    
""" 
#  Alternative solution for HI POINTS,HI REBOUNDS,HI ASSISTS:
    
espn = pd.melt(espn, id_vars=['DATE', 'OPPONENT','RESULT','W-L'], 
                 var_name='VARA', value_name='TEMPA')
temp = espn['TEMPA'].str.split(' ', n = 1, expand = True) 
espn['Player'] = temp[0] 
espn['Value'] = temp[1].astype('int64') 
espn.drop('TEMPA',axis=1,inplace = True)
espn = espn.pivot_table(index = ['DATE', 'OPPONENT','RESULT','W-L'],
                          columns='VARA',
                          values=['Player','Value'],
                          aggfunc={"Player":'max',"Value":'max'})

espn.columns=[' - '.join(str(s).strip() for s in col[::-1] if s) for col in espn.columns]
espn.reset_index(inplace = True)
""" 
    
espn['True Date'] = espn.apply(lambda x: x['DATE'] if x['DATE'].month<10 else x['DATE'] + relativedelta(years=-1),axis=1) 
espn['OPPONENT(clean)'] = espn.apply(lambda x: re.sub(r'^(@|vs)', '',x['OPPONENT']),axis=1) 
espn['Win or Loss'] = espn.apply(lambda x: x['RESULT'][0:1],axis=1)     
espn['Home or Away'] = espn.apply(lambda x: 'Away' if x['OPPONENT'][0:1]=='@' else 'Home',axis=1)     

final = espn[['OPPONENT(clean)','HI ASSISTS - Player','HI ASSISTS - Value','HI REBOUNDS - Player','HI REBOUNDS - Value', \
              'HI POINTS - Player','HI POINTS - Value','Win or Loss','Home or Away','True Date','OPPONENT','RESULT','W-L']]

File: 2019W04/preppindataw4.R
library(readxl)
library(dplyr)
library(tidyr)
library(lubridate)

espn <- read_excel("E:/PD - ESPN stats.xlsx", 
                   col_types = c("date", "text", "text", 
                                 "text", "text", "text", "text"))

clean_wl <- function(wl){
  sapply(wl,function(wl){
    if (grepl('.*-.*',wl)){
      return(wl)
    } else {
      td <- as.Date(as.numeric(wl),origin = "1899-12-30")
      return(paste0(as.character(day(td)),'-',as.character(month(td))))
    }  
  })
}

final <- espn %>%
  mutate('W-L' = clean_wl(`W-L`)) %>%
  gather(., key='vara', value='tempa', 'HI POINTS','HI REBOUNDS','HI ASSISTS') %>%
  separate(`tempa`,c('Player','Value')) %>%
  gather(., key='varb', value='tempb', 'Player','Value') %>%
  unite('varc', `vara`, `varb`, sep = ' - ') %>%
  spread(varc, tempb) %>%
  mutate_at(vars(matches(" - Value")),as.numeric) %>%
  mutate('True Date' = if_else(month(`DATE`)<10,`DATE`,`DATE` %m+% years(-1))) %>%
  mutate('OPPONENT(clean)' = gsub('^(@|vs)', '',`OPPONENT`)) %>%
  mutate('Win or Loss' = substr(`RESULT`, 1, 1)) %>%
  mutate('Home or Away' = if_else(grepl('^@',`OPPONENT`), 'Away', 'Home')) %>%
  select('OPPONENT(clean)','HI ASSISTS - Player','HI ASSISTS - Value','HI REBOUNDS - Player','HI REBOUNDS - Value',
         'HI POINTS - Player','HI POINTS - Value','Win or Loss','Home or Away','True Date','OPPONENT','RESULT','W-L')
   
View(final)